kmir/src/kmir/kdist/mir-semantics/rt/data.md

```diff
@@ -373,6 +373,91 @@ These helpers mark down, as we traverse the projection, what `Place` we are curr
   rule #decrementRef(TL) => #adjustRef(TL, -1)
 ```
 
+When a function returns a value, we need to decrement the reference count of the value.
+Because the current stack frame is destroyed after the return and will keep unchanged, 
+it's safe and necessary to traverse the projection until no current stack frame is needed.
+Otherwise, the semantics will lose the real meaning of the value to return.
+
+```k
+  syntax Value ::= #returnValue ( Value , List ) [function, total]
+  // `Value`: the value to return
+  // `List` : locals of the current stack frame
+  // --------------------------------------------------------
+  rule #returnValue(Reference(_, _, _, _) #as VAL, LOCALS)
+    => #decrementRef(#traverseLocalProjection(VAL, LOCALS, .List))
+  rule #returnValue(PtrLocal(_, _, _, _) #as VAL, LOCALS)
+    => #decrementRef(#traverseLocalProjection(VAL, LOCALS, .List))
+  rule #returnValue(TL, _) => #decrementRef(TL) [owise]
+
+  syntax Value ::= #traverseLocalProjection ( Value , List , List ) [function, total]
+  // `Value`: the value to traverse
+  // `List` : locals of the current stack frame
+  // `List` : projections to traverse
+  // --------------------------------------------------------
```

Comment:
My suggested approach was to avoid having the `Deref` in the stored projections inside a `Reference` or `PtrLocal` place , while this code tries to remove them only when returning the function. Your approach may be better because we avoid the work when we don't have to do it.
Two more comments: 
1. It is a bit hard to follow what exactly this function achieves...but there is some duplication between this function and `#traverseProjection`. We could try to use that one.
2. The "List" of projections is actually a list of lists, the outer list being a K list (untyped) and the inner one being  of sort `ProjectionElems`, which is a syntactic list.